R/stackingutilities.R

```diff
@@ -492,9 +492,241 @@ getMultiOutcomeNames <- function(data)
+    warning("NETs are removed from this analysis unless all their source values are mutually exclusive ",
+            "to other codes. The ", variable.type.and, " have ", net.txt, net.name.txt, " that contains ",
+            "source values that partially overlap with other codes. Consequently, ", net.txt.2, " not used in ",
+            "the analysis. If you wish ", net.txt.3, " to be used in the analysis then please modify the ",
+            variable.type.or, " via the Table view options appropriately.")
+}
+
+#' Stacks several text variables and an existing categorization 
+#' in a consisten manner for text analysis functions.
+#' 
+#' @param text A data frame containing one or more character vectors. Typically
+#' these start their life as variables in a Text - Multi variable set in Displayr
+#' or are several related Text variable sets.
+#' @param existing.categorization A data frame containing the corresponding
+#' categories. This could be a data frame of factor variables (a single-response 
+#' categorization), or a data frame of binary numeric variables (a multiple-response
+#' categorization). This can be NULL if no existing categorization is to be used
+#' in the analysis.
+#' @param subset A boolean vector indicating the subset of cases that the user
+#' wishes to include later in a classification of the text. Usually comes from QFilter.
+#' This will be expanded to align with the stacked data.
+#' @param weights A numeric vector of case weights for later us in the analysis.
+#' This will be expanded to align with the stacked data.
+#' @importFrom utils stack
+#' @importFrom flipFormat TidyLabels ExtractCommonPrefix
+#' @export
+StackTextAndCategorization <- function(text, existing.categorization = NULL, subset = TRUE, weights = NULL) {
+    
+    # One text variable, nothing to stack
+    if (is.list(text) && length(text) == 1) {
+        if (!is.null(existing.categorization) && 
+            ! attr(existing.categorization, "questiontype") %in% c("PickOne", "PickAny"))
+            stop("The existing categorization should be a Nominal/Ordinal or Binary - Multi variable set")
+        return(list(text = text[[1]],
+                    existing.categorization = existing.categorization,
+                    subset = subset,
+                    weights = weights))
+    }
+
+    if (!is.null(existing.categorization) && 
+        ! attr(existing.categorization, "questiontype") %in% c("PickOneMulti", "PickAnyGrid"))
+        stop("The existing categorization should be a Nominal/Ordinal - Multi or Binary - Grid variable set")
+
+    if (!is.data.frame(text)) {
+        text <- as.data.frame(text, optional = TRUE)
+        text.names <- TidyLabels(colnames(text))
+        colnames(text) <- text.names
+        attr(text, "codeframe") <- fakeCodeFrame(text.names)
+        attr(text, "questiontype") <- "TextMulti"
+        question.names <- vapply(text, FUN = function(x) attr(x, "question"), FUN.VALUE = character(1))
```

Comment:
Instead of defining the anonymous function, can use `vapply` but supply additional args in the final `...` arg.

```suggestion
        question.names <- vapply(text, FUN = attr, FUN.VALUE = character(1L), which = "question")
```